### PR TITLE
adding concurrency for calls to compileSassFile

### PR DESCRIPTION
**Pull request by 2cflor**
Adding 4 concurrent calls per cpu to compileSassFile, to reduce build times.

### CODE REVIEW

File: lib/broccoli_sass_compiler.js [hunk 1]
```diff
@@ -273,12 +274,23 @@ BroccoliSassCompiler.prototype.logCompilationFailure = function(details, error)
...
+  // throttle number of calls to compileSassFile
+  var numConcurrentCalls = os.cpus().length;
+  var promiseChunks = [];
+  for (var index = 0; index < promises.length; index += numConcurrentCalls) {
+    var chunk = promises.slice(index, index + numConcurrentCalls);
+    promiseChunks.push(chunk);
+  }
+ return promiseChunks.reduce(function(promise, chunk) {
+   return promise.then(function() {
+     return RSVP.all(chunk);
```
Comment (stefanpenner) [inline lib/broccoli_sass_compiler.js]: If one of these fails, we will not wait for everything to complete before moving on. That will most likely cause issues.

I raised similar concerns [here](https://github.com/stefanpenner/broccoli-persistent-filter/pull/111) I believe @mikrostew came up with a solution that mitigates the issues. I believe you guys are on the same team, if so you should sync up.

Comment (2cflor) [inline lib/broccoli_sass_compiler.js]: Thank you, I am checking in with Michael about this and will give it a try tonight.

File: lib/broccoli_sass_compiler.js [hunk 2]
```diff
@@ -288,7 +288,7 @@ BroccoliSassCompiler.prototype.compileTree = function(srcPath, files, destDir) {
   }
  return promiseChunks.reduce(function(promise, chunk) {
    return promise.then(function() {
-     return RSVP.all(chunk);
+     return RSVP.allSettled(chunk);
```
Comment (stefanpenner) [inline lib/broccoli_sass_compiler.js]: In addition to this, you will also need to "handle" the errors in the fulfillment. What i mean by this is:

```
RSVP.allSettled([]).then(settled => {
  settled.forEach(item => {
    if (item.state === 'rejected') {
      throw item.reason; // this way if something does fail, we wait until the concurrent work is done, and then throw.
    }
  });
})
```

Comment (2cflor) [inline lib/broccoli_sass_compiler.js]: Ah, gotcha. Thank you.

**Comment by 2cflor** [comment]
I added a variable for number of calls per core.
I made some experiments and found the lowest compile time is when we call sassCompileFile() 4 times per core. This seems to keep both the libsass parallelism and node concurrency busy enough, without slowing down from hitting the CPU ceiling.

**Comment by stefanpenner** [comment]
> I made some experiments and found the lowest compile time is when we call sassCompileFile() 4 times per core. This seems to keep both the libsass parallelism and node concurrency busy enough, without slowing down from hitting the CPU ceiling.

Interesting, that is a tad surprising. But maybe we are benefiting from some hyper-threading or some IO parallelism.  We can always tune it more then the future.

Should we make those constants available as environment variables? such as `process.env.PER_CPU_CONCURRENCY` or something? If you don't think we need to, don't worry about it

**Comment by 2cflor** [comment]
> Should we make those constants available as environment variables? such as process.env.PER_CPU_CONCURRENCY or something? If you don't think we need to, don't worry about it
I don't think we need it for our use case. Maybe other users might find it helpful? What do you think?

**Comment by stefanpenner** [comment]
> I don't think we need it for our use case. Maybe other users might find it helpful? What do you think?

we can always add it in the future if someone wants it.

**Comment by chriseppstein** [comment]
Thanks! Released as version `2.4.2`

**Comment by stefanpenner** [comment]
@chriseppstein although up to you. I would recommend changes like this also get corresponding test coverage.

**Comment by chriseppstein** [comment]
That would be good

**Comment by stefanpenner** [comment]
@2cflor do you mind adding some additional tests covering this PR, but in a new PR. If you have questions on how/why/what to test, feel free to reach out :)

**Pull request merged**